Add suspendables to throttler in consistent state

matthias-wende-sociomantic opened the pull request:
This change ensures that the suspend state of an ISuspendable instance
when added to the throttler is consistent with the throttler state.

This patch somehow also works around the the following problem.

If a RequestSuspender instance is reused by a new Request it'll set the suspend_requested flag to false.

https://github.com/sociomantic-tsunami/dhtproto/blob/v14.x.x/src/dhtproto/client/legacy/internal/request/model/IBulkGetRequest.d#L133
https://github.com/sociomantic-tsunami/swarm/blob/v5.x.x/src/swarm/client/request/helper/RequestSuspender.d#L148

When adding this instance to the throttlers list of suspendables it will be already present and therefore neither appended to the list of suspendables nor suspended.

https://github.com/sociomantic-tsunami/ocean/blob/v4.x.x/src/ocean/io/model/ISuspendableThrottler.d#L77

Now if at the same time the application (i.e. the applications throttler) is in suspended state, the request will be started in a non suspended state and will not get suspended at least until the throttler resumes and suspends again.

https://github.com/sociomantic-tsunami/ocean/blob/v4.x.x/src/ocean/io/model/ISuspendableThrottler.d#L139

**Note** that that doesn't address the issue that a suspendable is kept in the list of suspendables when a request finishes (compere also the comment below). The possibility for clients to unregister suspendables from the Throttller when a request finishes will be addressed later.

matthias-wende-sociomantic committed:
Add suspendables to throttler in consistent state

This change ensures that the suspend state of an ISuspendable instance
when added to the throttler is consistent with the throttler state.

daniel-zullo commented on src/ocean/io/model/ISuspendableThrottler.d:
```diff
@@ -77,10 +77,11 @@ abstract public class ISuspendableThrottler
+
+        if (this.suspended_ && !s.suspended())
+        {
+            s.suspend();
```
should it be suspended if it isn't added to the `suspendables`? Anyway if there is any reason to do so then the documentation of the method should be updated.

matthias-wende-sociomantic replied:
Yes that's the whole point. Recycled instances are in the list and they should be suspended.

matthias-wende-sociomantic replied:
And good point. I updated the doc accordingly ;).

daniel-zullo commented:
If this is a bug fix then it would be good to have/add a unit test

matthias-wende-sociomantic committed:
Add unittest for addSuspendable method

This test checks if an instance of ISuspendable that
is added to the throttler has the same suspend state as
the throttler.

matthias-wende-sociomantic commented:
> If this is a bug fix then it would be good to have/add a unit test

Yeah, thanks for pointing that out. I've added one for this particular case.

gavin-norman-sociomantic commented:
Hmm wasn't the idea originally to add a method allowing the user to remove a suspendable from the throttler?

matthias-wende-sociomantic commented:
Yes, that was the old plan before this new one :).
I favour this solution for several reasons.
- It fixes the bug as well.
- It is much simpler (i.e. no changes in swarm and dhtproto).
- It fixes the bug without changing clients.

At the same time the original idea introduced some new regression to the client and I have ~no~ intension to track this down ~if this patch works~ _later_.

gavin-norman-sociomantic commented:
You describe this situation:
1. A suspendable is registered with the throttler.
2. The suspendable ceases to be valid (e.g. the request it represented finished) and is recycled into a pool of suspendables. It is not unregistered from the throttler.
3. The suspendable is acquired from the pool and reinitialised.
4. The suspendable is registered with the throttler. As it is already registered, some special behaviour is required.

What I'm concerned about is this: what happens if, after step 2, the throttler tries to suspend? It would be calling a method of the recycled suspendable, which would do...? The fact that invalid suspendables can (must, currently) stay registered with the throttler seems like a bug to me, not something that we should patch up to make work.

What do you think?

gavin-norman-sociomantic commented:
Note that a recycled invalid suspendable is only one case. What if the invalid suspendable has been garbage collected, say?

matthias-wende-sociomantic commented:
Yes that's a valid point a should be addressed as well. My description of this patch was somehow misleading. I've updated the PR and commit descriptions to what this patch is actually about.

gavin-norman-sociomantic commented on src/ocean/io/model/ISuspendableThrottler.d:
```diff
@@ -91,6 +91,64 @@ abstract public class ISuspendableThrottler
+        suspendable_throttler.suspendAll();
+
+        // add a resumed ISuspendable to a suspended throttler
+        // note that the instance is "recycled"
```
This one comment probably isn't needed any more.

gavin-norman-sociomantic commented on src/ocean/io/model/ISuspendableThrottler.d:
```diff
@@ -91,6 +91,64 @@ abstract public class ISuspendableThrottler
+        suspendable.suspend();
+        suspendable_throttler.addSuspendable(suspendable);
+
+        test!("==")(suspendable.suspended(), false, "Unit test failed, " ~
```
Generally, I don't think you need to write "Unit test failed". It's clear from the context.

gavin-norman-sociomantic commented on src/ocean/io/model/ISuspendableThrottler.d:
```diff
@@ -91,6 +91,64 @@ abstract public class ISuspendableThrottler
+        suspendable.resume();
+        suspendable_throttler.addSuspendable(suspendable);
+
+        test!("==")(suspendable.suspended(), true, "Unit test failed, " ~
```
Generally, I don't think you need to write "Unit test failed". It's clear from the context.

gavin-norman-sociomantic reviewed (commented):
Yep, the new explanation makes much more sense, outside of the context of the bug you were investigating.

gavin-norman-sociomantic commented:
It seems like this should be a patch release.

matthias-wende-sociomantic commented:
Updated, base branch changed + rebased on top of new branch, accidentally deleted branch, recovered the branch and finally reopened the PR :).

daniel-zullo commented on src/ocean/io/model/ISuspendableThrottler.d:
```diff
@@ -77,13 +79,75 @@ abstract public class ISuspendableThrottler
+        else if (!this.suspended_ && s.suspended())
+        {
+            s.resume();
+        }
```
Why make it simple if we could make it complicated?

```D
        if (this.suspended_ != s.suspended())
        {
            if (this.suspended_)
                s.suspend();
            else
                s.resume();
        }
```

gavin-norman-sociomantic replied:
This PR has been merged, but that's a good point.

matthias-wende-sociomantic replied:
Yep, that's looks indeed better.

daniel-zullo commented on src/ocean/io/model/ISuspendableThrottler.d:
```diff
@@ -77,13 +79,75 @@ abstract public class ISuspendableThrottler
+        }
+    }
+
+    version (UnitTest){
```
the format/style is wrong

daniel-zullo commented on src/ocean/io/model/ISuspendableThrottler.d:
```diff
@@ -77,13 +79,75 @@ abstract public class ISuspendableThrottler
+
+            public bool suspended ( )
+            {
+                return suspended_;
```
missing `this.`
